### PR TITLE
Address paper review comments

### PR DESCRIPTION
**Pull request by mbudiu-vmw**
This tries to address reviewer comments for the VLDB paper.
With these changes the paper spills a paragraph into the last page.
We'll shrink it when we know how much space we have for the camera ready.
If we get more space we may add more examples...

### CODE REVIEW

File: doc/vldb23/nested.tex [hunk 1]
```diff
@@ -168,9 +168,11 @@ \subsection{Cost of incremental recursive queries}
 once the non-incremental circuit reaches the fixed point, its output is constant, 
 and the derivative of corresponding value in the incrementalized circuit becomes 0.
 
-Moreover, the work performed by each operator in the incremental 
-circuit is asymptotically better than the non-incremental one.
-A detailed analysis can be found in our companion report~\cite{tr}.
+Moreover, the work performed by each operator in the incremental
+circuit is asymptotically less than the non-incremental
+one\footnote{Under the standard assumption enumerated
```
Comment (ryzhyk) [inline doc/vldb23/nested.tex:173]: assumptionS

File: doc/vldb23/relational.tex [hunk 2]
```diff
@@ -21,24 +21,24 @@ \section{IVM for the Relational Algebra}\label{sec:relational}
...
-Given a set $A$, we define \defined{\zrs}
-over $A$ as functions with \emph{finite support} from $A$ to $\Z$.
-These are functions $f: A \rightarrow \Z$ where
-$f(x) \not= 0$ for at most a finite number of values $x \in A$.
-We also write $\Z[A]$ for the type of \zrs with elements from $A$.
-Values in $\Z[A]$ can be thought of as key-value maps with
-keys in $A$ and values in $\Z$, justifying the array indexing notation.
-If $m \in \Z[A]$ we write $m[a]$ instead of $m(a)$, again using
-an indexing notation.
+\zrs generalize databases tables: think of a \zr as a table where each
```
Comment (ryzhyk) [inline doc/vldb23/relational.tex:24]: database(S)

File: doc/vldb23/relational.tex [hunk 3]
```diff
@@ -624,4 +623,35 @@ \subsection{Complexity of incremental circuits}\label{sec:complexity}
 data persistently.  The space complexity of operators such as $\inc{(\lift{\distinct})}$,
 $\inc{(\lift{\bowtie})}$, $\I$, and $\D$ is $O(|DB[t]|)$.
 
+\subsubsection{IVM query plans and optimality}
+
+We want to spell explcitly what Algorithm~\ref{algorithm-inc} actually
+peforms.
```
Comment (ryzhyk) [inline doc/vldb23/relational.tex:629]: typo

File: doc/vldb23/relational.tex [hunk 4]
```diff
@@ -516,26 +518,23 @@ \subsection{Incremental view maintenance}
...
- Most of the operators that appear in the circuits in
+Step (2) generates an equivalent circuit, with possibly fewer
+$\distinct$ operators.  Step (3) yields a circuit that consumes a
+\emph{stream} of complete database snapshots and outputs a stream of
+complete view snapshots. Step (4) yields a circuit that consumes a
+stream of \emph{database changes} and outputs a stream of \emph{view
+changes}; however, the internal operation of the circuit is
+non-incremental, as it rebuilds the complete database using
+integration operators.  Step (5) incrementalizes the circuit by
+replacing each primitive operators with its incremental version.
```
Comment (ryzhyk) [inline doc/vldb23/relational.tex:529]: typo

File: doc/vldb23/relational.tex [hunk 5]
```diff
@@ -624,4 +623,35 @@ \subsection{Complexity of incremental circuits}\label{sec:complexity}
 data persistently.  The space complexity of operators such as $\inc{(\lift{\distinct})}$,
 $\inc{(\lift{\bowtie})}$, $\I$, and $\D$ is $O(|DB[t]|)$.
 
+\subsubsection{IVM query plans and optimality}
+
+We want to spell explcitly what Algorithm~\ref{algorithm-inc} actually
+peforms.
+
+In relational algebra a query can be implemented by multiple plans,
+each with a different data-dependent cost\footnote{I.e., which plan is
```
Comment (ryzhyk) [inline doc/vldb23/relational.tex:632]: "I.e.," cannot start a sentence.

File: doc/vldb23/relational.tex [hunk 6]
```diff
@@ -624,4 +623,35 @@ \subsection{Complexity of incremental circuits}\label{sec:complexity}
...
+optimal may depend not only on the query, but also on the data.}.  The
+input of Algorithm~\ref{algorithm-inc} is not an abstract query, it is
```
Comment (ryzhyk) [inline doc/vldb23/relational.tex:634]: I would remove the "not an abstract query" part:

The input of Algorithm X is a non-incremental query plan produced by a query planner.

File: doc/vldb23/relational.tex [hunk 7]
```diff
@@ -624,4 +623,35 @@ \subsection{Complexity of incremental circuits}\label{sec:complexity}
...
+historical statistics, can be used to generate the query plan that is
+supplied to Algorithm~\ref{algorithm-inc}.  The question of optimality
+in the context of IVM plan is a much more difficult topic than
+optimization of ad-hoc queries, since the chosen IVM plan will execute
+for \emph{all future database updates}.
+
+Moreover, since incremental computations maintain internal state, it
+follows that incremental plans cannot be simply changed in-flight,
+like we can change ad-hoc queries based on current data statistics: a
+new plan has to ``learn'' the state produced by entire history of
```
Comment (ryzhyk) [inline doc/vldb23/relational.tex:652]: It may not be clear what "learn" means here.

**Review by ryzhyk — CHANGES_REQUESTED** [review]
Minor nits

File: README.md [hunk 8]
```diff
@@ -40,10 +40,13 @@ logic efficiently.
 
 The above objectives can only be achieved by building on a solid mathematical
 foundation.  The formal model that underpins our system, also called DBSP, is
-described in the accompanying paper:
+described in the accompanying paper (link points to an earlier version while
```
Comment (ryzhyk) [inline README.md:43]: I suggest we point the reference to this version, which is much better than the one on archive and is very close to final. Just add  a [preprint] label to the PDF.

Comment (mbudiu-vmw) [reply README.md:43]: Amazing how many mistakes one can do in a small edit.
Yes, but I don't want to commit the pdf yet. Once we have a camera-ready we'll switch.
I am secretly hoping they will give us extra space so we can slip one more example.
If not, we'll have to do some work to squeeze this version to fit the extra paragraphs I added.

Comment (ryzhyk) [reply README.md:43]: Your call, but we keep hearing that people tried to read the arxiv paper, and didn't fully understand it.  I don't see any downside to pointing them to a much better write-up even if it's not final.

Comment (mbudiu-vmw) [reply README.md:43]: For papers having multiple versions is not as good. Let's just wait a few days until we get instructions for the camera-ready and we'll package a final version. We could perhaps create a longer version with additional examples and less math if you want that.

**Pull request merged**